src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1145,17 +1265,28 @@ public Solution WithDocumentSourceCodeKind(DocumentId documentId, SourceCodeKind
         /// </summary>
         public Solution WithDocumentTextLoader(DocumentId documentId, TextLoader loader, PreservationMode mode)
         {
-            return WithDocumentTextLoader(documentId, loader, text: null, mode: mode);
-        }
+            CheckContainsDocument(documentId);
 
-        internal Solution WithDocumentTextLoader(DocumentId documentId, TextLoader loader, SourceText? text, PreservationMode mode)
-        {
-            var newState = _state.WithDocumentTextLoader(documentId, loader, text, mode);
-            if (newState == _state)
+            if (loader == null)
             {
-                return this;
+                throw new ArgumentNullException(nameof(loader));
+            }
+
+            if (!mode.IsValid())
+            {
+                throw new ArgumentOutOfRangeException(nameof(mode));
             }
 
+            return UpdateDocumentTextLoader(documentId, loader, text: null, mode: mode);
+        }
+
+        internal Solution UpdateDocumentTextLoader(DocumentId documentId, TextLoader loader, SourceText? text, PreservationMode mode)
+        {
+            var newState = _state.UpdateDocumentTextLoader(documentId, loader, text, mode);
+
+            // Note: state is currently not reused.
+            Debug.Assert(newState != _state);
```

Comment:
Why assert a _bad_ behavior?

Reply:
Or at least, can the comment be revised to clarify this isn't bad? I worry that if somebody fixed this they'd end up backing the change out because of this assert, and that's probably not what we'd want.

Reply:
I don't think the behavior is bad. It could be improved, for sure. I thought the comment above is sufficient to state that. I'll update the comment to be more clear.

If I just remove the assert and `UpdateDocumentTextLoader` reuses the instance we still return new Solution instance from this method instead of reusing it. The assert will let us know we should update the code.